src/router/actor.rs

```diff
@@ -0,0 +1,1088 @@
+    filter::Filter,
+    message::{RelayMessage, SubscriptionId},
+    nips::nip44,
+};
+use nostr_relay_pool::RelayPoolNotification;
+use serde::{Serialize, de::DeserializeOwned};
+use tokio::sync::{mpsc, oneshot};
+use tokio_stream::StreamExt;
+
+use crate::{
+    protocol::{LocalKeypair, model::event_kinds::SUBKEY_PROOF},
+    router::{
+        CleartextEvent, Conversation, ConversationError, ConversationMessage, NotificationStream,
+        PortalId, RelayNode, Response, channel::Channel,
+    },
+};
+
+type ConversationBox = Box<dyn Conversation + Send + Sync>;
+
+impl std::fmt::Debug for ConversationBox {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Conversation").finish()
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum MessageRouterActorError {
+    #[error("Channel error: {0}")]
+    Channel(#[from] Box<dyn std::error::Error + Send + Sync>),
+    #[error("Conversation error: {0}")]
+    Conversation(#[from] ConversationError),
+    #[error("Receiver error: {0}")]
+    Receiver(#[from] oneshot::error::RecvError),
+}
+
+#[derive(Debug)]
+pub enum MessageRouterActorMessage {
+    AddRelay(String, oneshot::Sender<Result<(), ConversationError>>),
+    RemoveRelay(String, oneshot::Sender<Result<(), ConversationError>>),
+    Shutdown(oneshot::Sender<Result<(), ConversationError>>),
+    Listen(oneshot::Sender<Result<(), ConversationError>>),
+    AddConversation(
+        ConversationBox,
+        oneshot::Sender<Result<PortalId, ConversationError>>,
+    ),
+    AddConversationWithRelays(
+        ConversationBox,
+        Vec<String>,
+        oneshot::Sender<Result<PortalId, ConversationError>>,
+    ),
+    SubscribeToServiceRequest(
+        PortalId,
+        oneshot::Sender<Result<NotificationStream<serde_json::Value>, ConversationError>>,
+    ),
+    AddAndSubscribe(
+        ConversationBox,
+        oneshot::Sender<Result<NotificationStream<serde_json::Value>, ConversationError>>,
+    ),
+    Ping(oneshot::Sender<()>),
+
+    /// This is used to handle relay pool notifications.
+    HandleRelayPoolNotification(RelayPoolNotification),
+
+    /// This is used from SDK to get the relays.
+    GetRelays(oneshot::Sender<Result<Vec<String>, ConversationError>>),
+}
+
+pub struct MessageRouterActor<C>
+where
+    C: Channel + Send + Sync + 'static,
+    C::Error: From<nostr::types::url::Error>,
+{
+    channel: Arc<C>,
+    keypair: LocalKeypair,
+    sender: mpsc::Sender<MessageRouterActorMessage>,
+}
+
+impl<C> MessageRouterActor<C>
+where
+    C: Channel + Send + Sync + 'static,
+    C::Error: From<nostr::types::url::Error>,
+{
+    pub fn new(channel: C, keypair: LocalKeypair) -> Self {
+        let keypair_clone = keypair.clone();
+        let channel = Arc::new(channel);
+
+        let (tx, mut rx) = mpsc::channel(4096);
+
+        let channel_clone = Arc::clone(&channel);
+        let tx_clone = tx.clone();
+        tokio::spawn(async move {
```

Comment:
Spawning tasks doesn't work if there isn't a runtime set.

It's ok to introduce this change but in the app, since the methods are called through uniffi and thus don't have a runtime, you need to wrap the calls that can eventually lead to spawning tasks (i think this is only happening in the constructor right?) in the async_utils::task::spawn thing.

You can see how I've done it for the calls to cashu since it's also spawning tasks internally in the PR #62